### PR TITLE
Fix fast sync transition block missed

### PR DESCRIPTION
**Pull request by asdacap**
- Fix trieexception caused by attempting to execute block when parent's state root is missing. Block is wrongly attempted due missed fast sync transition.
- In some case, state sync just execute twice and it would still synced fine.

## Changes

- Add unit test.
- Fix issue.
- Add check to prevent executing when parent's state root is missing.

#### What types of changes does your code introduce?

- [X] Bugfix (a non-breaking change that fixes an issue)

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

#### Notes on testing

_Optional. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

**Comment by LukaszRozmej** [comment]
Doesn't it duplicate #5114 ?

File: src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs [hunk 1]
```diff
@@ -618,8 +632,7 @@ private ProcessingBranch PrepareProcessingBranch(Block suggestedBlock, Processin
                 // we need to dig deeper to go all the way to the false (reorg boundary) head
                 // otherwise some nodes would be missing
                 bool notFoundTheBranchingPointYet = !_blockTree.IsMainChain(branchingPoint.Hash!);
-                bool notReachedTheReorgBoundary = (options & ProcessingOptions.Trace) == ProcessingOptions.Trace
-                && (branchingPoint.Number > (_blockTree.Head?.Header.Number ?? 0));
+                bool notReachedTheReorgBoundary = branchingPoint.Number > (_blockTree.Head?.Header.Number ?? 0);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:635]: Is this conscious change? This was just changed in https://github.com/NethermindEth/nethermind/pull/5090

File: src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs [hunk 2]
```diff
@@ -519,6 +519,20 @@ void DeleteInvalidBlocks(Keccak invalidBlockHash)
...
+                    BlockHeader? parentOfFirstBlock = _blockTree.FindHeader(blocksToProcess[0].ParentHash!, BlockTreeLookupOptions.None);
+                    if (parentOfFirstBlock == null)
+                    {
+                        throw new InvalidOperationException("Attempted to process a disconnected blockchain");
+                    }
+
+                    if (!_stateReader.HasStateForBlock(parentOfFirstBlock))
+                    {
+                        throw new InvalidOperationException("Attempted to process a blockchain without having starting state");
+                    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:534]: I expect that some other exception would have been thrown anyway?

Comment (asdacap) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:534]: Well yea, but TrieException. Not exactly clear whats wrong.

**Comment by asdacap** [comment]
Why does it duplicate that?

File: src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:635]: Yea, I think @OlegJakushkin said that @MarekM25 said that this is needed?

Comment (MarekM25) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:635]: @asdacap @OlegJakushkin @LukaszRozmej  just to be clear, I proposed this change as the temp hack to see if it resolves our memory problems with `trace_block`

Comment (MarekM25) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:635]: and this condition is wrong. I didn't want to check reachedTheReorgBoundry for traces. So like this:
` bool notReachedTheReorgBoundary = (options & ProcessingOptions.Trace) != ProcessingOptions.Trace
&& (branchingPoint.Number > (_blockTree.Head?.Header.Number ?? 0));`

not like this:
` bool notReachedTheReorgBoundary = (options & ProcessingOptions.Trace) == ProcessingOptions.Trace
&& (branchingPoint.Number > (_blockTree.Head?.Header.Number ?? 0));`

Comment (OlegJakushkin) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:635]: @LukaszRozmej This was not needed to fix trace block at all as there were other checks more, global. Will remove it in next PR on this topic later today.

**Review by MarekM25 — APPROVED** [review]

File: src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs (hunk 1 above)
Comment (kamilchodola) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:635]: @OlegJakushkin Was it improved on any other PR?

**Pull request merged**